docs/jotai/advanced-recipes/atom-creators.mdx

```diff
@@ -0,0 +1,92 @@
+---
+title: Atom creators
+nav: 5.2
+---
+
+## `atomWithToggle`
+
+> `atomWithToggle` creates a new atom with a boolean as initial state & a setter function to toggle it.
+
+This avoids the boilerplate of having to setup another atom just to update the state of the first.
+
+```ts
+import { WritableAtom, atom } from 'jotai'
+import { atomWithStorage } from 'jotai/utils'
+
+export function atomWithToggle(initialValue?: boolean): WritableAtom<boolean, boolean | undefined> {
+  const anAtom = atom(initialValue, (get, set, nextValue?: boolean) => {
+    const update = nextValue ?? !get(anAtom)
+    set(anAtom, update)
+  })
+
+  return anAtom
+}
+```
+
+An optional initial state can be provided as first argument.
+
+The setter function can have an optional arg to force a particular state (if you want to make a setActive function out of it for example).
+
+Here is how it's used.
+
+```ts
+import { atomWithToggle } from 'XXX'
+
+// will have an initial value set to true
+const isActiveAtom = atomWithToggle(true)
+```
+
+And in a component :
+
+```tsx
+const Toggle = () => {
+  const [isActive, toggle] = useAtom(isActiveAtom)
+
+  return (
+    <>
+      <button onClick={() => toggle()}>isActive: {isActive ? 'yes' : 'no'}</button>
+      <button onClick={() => toggle(true)}>force true</button>
+      <button onClick={() => toggle(false)}>force false</button>
+    </>
+  )
+}
+```
+
+## `atomWithToggleAndStorage`
+
+> `atomWithToggleAndStorage` does the same as [`atomWithToggle`](#atomWithToggle) but also persist the state anytime it changes in given storage using [`atomWithStorage`](/jotai/api/utils#atomWithStorage).
+
+Here is the source :
+
+```ts
+import { WritableAtom, atom } from 'jotai'
+import { atomWithStorage } from 'jotai/utils'
+
+export function atomWithToggleAndStorage(
+  key: string,
+  initialValue?: boolean,
+  storage?: any
```

Comment:
```suggestion
  storage?: Parameters<typeof atomWithStorage>[2]
```

One more thing. Does this work?

Reply:
Partially. It's fine for the `storage` arg but then the returns would require a type casting since storage gets inferred as `Storage<unknown>
https://imgur.com/a/Z4sjYAY

Reply:
Oh, because it's generics. That's unfortunate. We'd need a better workaround in the future.